### DISCUSSION TITLE
Save high scores

### DISCUSSION THREAD

**Discussion by tlainez** [EmulatorJS/EmulatorJS#707] [General]
Hi,

I'm trying to build a site using emulatorJS and I need to save the high scores for each game in my server. I'm using roms with the default arcade core. I guess I need to specify the path for the highscore.dat, probably with some core option, but I'm not sure how. Is it posible? Do I need to host the core files to achieve this? Do I need to deal with some Retroarch configuration?
I'm quite newbie, so sorry if I'm doing basic questions, but I didn't find any info regarding this 😅
Thanks and congratulations to the creators & contributors of this amazing project!

Toni

**Comment by Underdog-01** [comment]
I already developed a platform that organizes & plays them using this plugin as well & would really like this feature for games that have a score.
Unfortunately it's only possible if someone modified each game to initiate some sort of ajax call & likely better to have some sort of save button (or an auto save in some circumstances). I think this idea is best mentioned on a hombebrew development site or maybe on a game devs GitHub page.  Also I think that someone out to add this to existing games somehow need to do it with multiple ROM editing platforms & somehow add it using RAW hex code which looks impossible & insane to me.

**Comment by michael-j-green** [comment]
MAME (and probably other cores too) have support for a file called HISCORE.DAT. This file describes how MAME should extract the high scores from games.

I haven't tried it out yet (I plan on playing with it in the next month or so), but the containing web page should upload the HISCORE.DAT file into the EJS virtual file system in the correct place, and that should be it. MAME would then start outputting files containing the scores named for the ROM that was loaded (the MAME documentation has all this documented).

The containing web page would periodically monitor the virtual file system for the presence of these files and download/act on them as needed. It'd need to be a poll since to my knowledge, there's no way to generate an event when these files are saved to the virtual file system.

**Comment by Underdog-01** [comment]
If I understand correctly.. it puts the score & game file name (for identity) into memory & then I suppose one would have to use setInterval() to continuously check if those were set. At that point I can have it do the Ajax Post call to save the score server side.  If something like this is possible it will be a great addition!

Just some thoughts.. 
keep in mind any games that save score like that likely need to have the save point, rewind, speed, etc. disabled unless someone wants people to cheat on the scores of hosted games that save a score. That toolbar of options is in the /data/emulator.js file so if one hosts something like this that file (if altered to facilitate custom edits) can't be loaded via the CDN unless the devs here change it up to allow optional disabling of things on that toolbar/menu from the index ( ie. via localStorage() or sessionStorage() )

**Comment by michael-j-green** [comment]
In my mind, all of this is outside of the scope of the EJS project, and would be the responsibility of the containing app. At most, EJS could provide API endpoints that the containing page could call.

For auto save, the containing page can call the save page function using a SetInterval to call a save state command and then pulling the state file out (I don’t yet know how to do this).

For scores, it would be largely the same… upload the hiscores.dat file when starting the core, and then poll for the presence of output files. I believe you could also upload the last high score table for persistence. Again, I don’t have any further details as I haven’t looked into the details yet.

**Comment by fedekrum** [comment]
Hi. I would love an api call or something to save the hi score. Any of the devs can say if this is at their sight ?

**Reply by S-Coding23** [reply]
I recently integrated RA on a separate project of mine. I might be able to add some kind of API integration here for high scores, but its going to be a little ways out if I can.
